### PR TITLE
fetched_balance_block_number instead of fetched_balance_at

### LINKED ISSUE #243: Addresses should fetch balance at a specific block, not time, and record that block
# Motivation

When the `Explorer.Indexer.AddressBalanceFetcher` fetches an `t:Explorer.Chain.Address.t/0` `fetched_balance`

https://github.com/poanetwork/poa-explorer/blob/4526874a8ceb1e5152f6b47001a20c9a5ba63c05/apps/explorer/lib/explorer/indexer/address_balance_fetcher.ex#L48-L50

it calls `EthereumJSONRPC.fetch_balances_by_hash/1`

https://github.com/poanetwork/poa-explorer/blob/4526874a8ceb1e5152f6b47001a20c9a5ba63c05/apps/ethereum_jsonrpc/lib/ethereum_jsonrpc.ex#L94-L111

which uses `latest` for the block number, which means the balance is as up-to-date as possible, but the problem is that we don't actually know which block number the balance is for then and instead we record a timestamp in `balance_fetched_at`.

`balance_fetched_at` kinda gives an idea of how out-of-date a balance may be, but it is difficult to correlate with the blocks, transactions, and internal transactions that may have changed the balance since the time is when we got the value, not the timestamp of block that the value is for.

# Proposed Changes

- [ ]  Instead of of using `latest` for the block parameter, feed in the desire block number from whatever is generating the address hash added to the `Explorer.Indexer.AddressBalanceFetcher`
- [ ] Replace `addresses.balance_fetched_at` with `addresses.balance_block_number`

## Follow-on Changes

Once we use `balance_block_number`, we can change `Explorer.Indexer.AddressBalanceFetcher` to to look ahead in the queue to see if there's a new block number queued and use the newest block number for the address.

Alternatively, we can move balances to a separate table and record the fetched balance at each block number.  We can combine this with the value generated in block mining/validating, transactions, and internal transactions to show a "cheque book" view of the address show which parts have been validated by the balance fetcher vs what we're showing from the indexed tables.  This will point out when our value calculations differ from the API.

### PR DESCRIPTION
**Pull request by KronicDeth**
Fixes #243

Instead of getting the address balance at the latest block and recording the timestamp of the insert, which is only loosely correlated with even the latest block's timestamp on-chain, use the block number for the last time an address was used in the indexed address foreign keys in the
chain:

* blocks
  * miner_hash
* internal_transactions
  * created_contract_address_hash
  * from_address_hash
  * to_address_hash
* logs
  * address_hash
* transactions
  * from_address_hash
  * to_address_hash

To gather `fetched_balance_block_number` on reboot, `Explorer.Chain.stream_unfetched_addresses` queries across the above columns for the `MAX(blocks.number)`, which is used in `Explorer.Indexer.AddressBalanceFetcher.init/2`.

During indexing, `Explorer.Indexer.AddressBalanceFetcher.async_fetch_balances/1` now requires a list of `%{block_number: Block.block_number(), hash: Hash.Truncated.t()}` instead of list of `Hash.Truncated.t()`.

Having a `block_number` available for all importable entities's addresses required changing `Explorer.Indexer.AddressExtraction` to extract the `block_number` too.  Because it was difficult to determine which parts of the indexer weren't supplying a `block_number` or `nil` block number, `Explorer.Indexer.AddressExtraction` was rewritten to error when no extract format matches for the list of extract formats for each entity.  These errors allowed to track where in block_number needed
to be passed along to prevent a missing block number except in the case of pending transaction where block number is expected to be `nil`.

The use of `insert` in the factories was making tests that show the count not make sense, so the factories were switched to `build` because it means only the exact count is created when a default value is overridden.  This does, unfortunately, mean that `params_for(factory)` no longer generates foreign keys and they must be set explicitly OR the association passed instead of the foreign key.

**NOTE**: This is a database drop and reindex change.

### CODE REVIEW

File: apps/explorer/lib/explorer/indexer/address_extraction.ex [hunk 1]
```diff
@@ -36,64 +40,273 @@ defmodule Explorer.Indexer.AddressExtraction do
...
-    |> extract_addresses_from_collection(@transactions_address_maps)
-    |> List.flatten()
-    |> merge_addresses()
-  end
+  defstruct pending: false
 
   @doc """
   Extract addresses from block, internal transaction, transaction, and log parameters.
+
+  Blocks have their `miner_hash` extracted.
```
Comment (igorffs) [inline apps/explorer/lib/explorer/indexer/address_extraction.ex]: These examples seem more suitable as tests than examples in the doc tests. The very specific test scenarios (that aren't explicit on the tests) and some already explained behaviors (in the module doc) gave that perception on me. 

I think we could move that to the tests and keep some of the new behavior explanation (about the raise) to the module doc. wdyt?

File: apps/explorer/lib/explorer/indexer/address_extraction.ex [hunk 2]
```diff
@@ -102,59 +315,146 @@ defmodule Explorer.Indexer.AddressExtraction do
...
+      end)
+
+    if count == 0 do
+      formatted_extract_format_list =
+        extract_format_list
+        |> Enum.with_index()
+        |> Enum.map(fn {extract_format, index} -> "#{index + 1}. #{inspect(extract_format)}}" end)
+        |> Enum.join("\n")
+
+      raise ArgumentError,
```
Comment (igorffs) [inline apps/explorer/lib/explorer/indexer/address_extraction.ex]: > Because it was difficult to determine which parts of the indexer weren't supplying a block_number or nil block number, Explorer.Indexer.AddressExtraction was rewritten to error when no extract format matches for the list of extract formats for each entity.

Can you explain a little more about the problem you're facing that the exception helps us? For instance are you mentioning `BlockFetcher`, `InternalTransactionFetcher` and `AddressBalanceFetcher` when you say "parts of the indexer"?

File: apps/explorer/lib/explorer/chain.ex [hunk 3]
```diff
@@ -179,17 +179,121 @@ defmodule Explorer.Chain do
...
+        ]}
+      iex> {:ok, changed_address} = Explorer.Chain.hash_to_address(hash)
+      iex> changed_address.fetched_balance
+      %Explorer.Chain.Wei{value: Decimal.new(1)}
+      iex> changed_address.fetched_balance_block_number
+      3
+
+  `t:Explorer.Chain.Address.t/0`s when first imported do not have a balance, in such a case, the balance updates win.
+
+      iex> address = insert(:address, hash: "0x8bf38d4764929064f2d4d3a56520a76ab3df415b")
```
Comment (amandasposito) [inline apps/explorer/lib/explorer/chain.ex]: WDYT of a test for this method? This way we could leave the `doctest` simpler and not worry about this setups here.

File: apps/explorer/lib/explorer/chain.ex [hunk 4]
```diff
@@ -1092,44 +1181,309 @@ defmodule Explorer.Chain do
...
-      iex> [first_address_hash, second_address_hash] = 2 |> insert_list(:address) |> Enum.map(& &1.hash)
-      iex> {:ok, address_hash_set} = Explorer.Chain.stream_unfetched_addresses([:hash],
-      ...>   MapSet.new([]),
-      ...>   fn %Explorer.Chain.Address{hash: hash}, acc ->
-      ...>     MapSet.put(acc, hash)
-      ...>   end
+  An `t:Explorer.Chain.Address.t/0` `hash` can be used as an `t:Explorer.Chain.Block.t/0` `miner_hash`.
+
+      iex> {:ok, miner_hash} = Explorer.Chain.string_to_address_hash("0xe8ddc5c7a2d2f0d7a9798459c0104fdf5e987aca")
+      iex> miner = insert(:address, hash: miner_hash)
```
Comment (amandasposito) [inline apps/explorer/lib/explorer/chain.ex]: same here,  WDYT of moving this to a test method?

File: apps/explorer/lib/explorer/indexer/address_extraction.ex (hunk 2 above)
Comment (KronicDeth) [reply apps/explorer/lib/explorer/indexer/address_extraction.ex]: The code was errorring at the `Chain.import_blocks` call or `EthereumJSONRPC.fetch_balances` depending on the point in development of the branch, but determining who was sending the bad address params that missed the `fetched_balance_block_number` was difficult because there are 3 paths that can supply `EthereumJSONRPC.fetch_balances`.  By erroring here, it was easier to trace the source of the incomplete address data.  This will also make it easier to see that we're missing the block number if a new fetcher is added that supplies addresses.  Without this error, those new addresses would follow through `AddressBalanceFetcher.async_fetch_balances`, where they'd be buffered with params from other addresses.

`async_fetch_balances` now checks for the proper format through `address_fields_to_params`, but it may be called in such as manner that tail call optimization will have eliminated the source the address fields sufficient to find the path that supplied the incomplete fields.  I also developed the exception system first.

I find the large exception message, which shows what patterns we were trying to match easier to read for what is missing than having a `FunctionClauseError` on `address_fields_to_params` as that won't show the full fields for the original entity, such as a transaction or internal transaction.

File: apps/explorer/lib/explorer/indexer/address_extraction.ex (hunk 1 above)
Comment (KronicDeth) [reply apps/explorer/lib/explorer/indexer/address_extraction.ex]: I wasn't going to change it in this PR, but I was going to bring up that I don't think we should document `@entity_to_extract_format_list` in the `@moduledoc`.  `@entity_to_extract_format_list` is not a register module attribute, so someone reading the docs, can't retrieve its value.  Instead, since `@entity_to_extract_format_list` should be documented in a normal comment using `#` above the module attribute since the module attribute is an implementation detail.  If that's the case the duplicates from the `@moduledoc` would be eliminated.

As for why to include `blocks` `miner_hash` when the example is in the `@moduledoc`?  Well, the `@moduledoc` explains how it works, but it doesn't actually show an example of input and output.  The `@doc` here shows the input and output of the function without regard to the module attribute, which is an implementation detail.  This way the docs show what happens with each type of entity so anyone reading just the docs and not the source can see the behavior for each entity.

If I can make something work as a doctest I will favor doing that over a test in a test module as it allows me to use more prose to explain the scenario than is possible with a test's description, which is restricted to the 256 unicode code point limit of the resulting function name.

File: apps/explorer/lib/explorer/chain.ex (hunk 3 above)
Comment (KronicDeth) [reply apps/explorer/lib/explorer/chain.ex]: I'd prefer to keep a doctest demonstrating the above prose than have the prose by itself. 

We use `insert` in doctests through this module and they work fine because doctests are compiled to normal `test` equivalent functions that use the same `import` and `setup`, so the database is cleaned up using the sandbox system.  Do you not like that I'm hard-coding a specific, arbitrary address?

File: apps/explorer/lib/explorer/chain.ex (hunk 4 above)
Comment (KronicDeth) [reply apps/explorer/lib/explorer/chain.ex]: Again, we're already using `insert` in other doctests in the module.  Is your objection the use of a specific address?  That is needed so the the printing of `address_fields_list` can be precomputed to show it works.

Is it the throw-way `Explorer.Chain.stream_unfetched_addresses(` call?  I can understand that since it's not really how the code is called, but not including this examples means moving the other doctests to also be normal tests then.

If you want me to not include tests that I can make work as doctests as doctests, but based on there being too much setup could you explain what criteria you want us to use for how much setup is too much setup?

I like doctest because anyone can copy them in the docs to try out the code, so I go with "if I can make it work as a doctest, do a doctest".  The doctests also double-check that the setup isn't too onerous, which guides making the API better as I have to explain how the function explains in full in prose.  Doing it as `test` in test module, for me it's more walking the decision tree of the branches and just making sure I have coverage without necessarily thinking out if it is easy to explain in prose.

File: apps/explorer/lib/explorer/chain.ex (hunk 3 above)
Comment (KronicDeth) [reply apps/explorer/lib/explorer/chain.ex]: Discussion continued in https://github.com/poanetwork/poa-explorer/issues/290.  Paused for this PR.

File: apps/explorer/lib/explorer/chain.ex (hunk 4 above)
Comment (KronicDeth) [reply apps/explorer/lib/explorer/chain.ex]: Discussion continued in https://github.com/poanetwork/poa-explorer/issues/290.  Paused for this PR.

File: apps/explorer/lib/explorer/indexer/address_extraction.ex (hunk 1 above)
Comment (KronicDeth) [reply apps/explorer/lib/explorer/indexer/address_extraction.ex]: Discussion continued in https://github.com/poanetwork/poa-explorer/issues/290.  Paused for this PR.

File: apps/explorer/lib/explorer/indexer/address_extraction.ex (hunk 2 above)
Comment (igorffs) [reply apps/explorer/lib/explorer/indexer/address_extraction.ex]: Got it @KronicDeth, thank your for the explanation. ❤️ ✨  

If I understood correctly we'll have this problem only during development, right? If that's the case I think we shouldn't have an exception for that. Testing what is the expected outcome when a function receives some sort of parameters should give proper feedback for it. 

We had a test that tested the behavior of [what would happen while using a List in the `@entity_to_extract_format_list`](https://github.com/poanetwork/poa-explorer/pull/283/files#diff-c561e99f50823e571ac2e38428be15afL143) and another that would [make sure the returns would go as expected](https://github.com/poanetwork/poa-explorer/pull/283/files#diff-c561e99f50823e571ac2e38428be15afL31). Maybe they should need to be more be more specific, for example:

In this scenario, I'd just add some test like:

```elixir
    test "blocks should return x" do
      response = Explorer.Indexer.AddressExtraction.extract_addresses(
        %{
          blocks: [
            %{
              miner_hash: "some hash"
              number: 34
            }
          ]
         }
      )

      assert response == [%{hash: "some hash", fetched_balance_block_number: 34 }]
    end
```

If that's a real scenario and a problem that happens when we're integrated with the network I'd probably consider a different option like logging. 

wdyt?

Comment (KronicDeth) [reply apps/explorer/lib/explorer/indexer/address_extraction.ex]: The bug only showed up in real-world usage though, when I was running tests in AWS, so I want the exception there to catch it the next time it happens in production.  If it was something I caught during development, I'd be more confident saying the test addresses it.

Comment (KronicDeth) [reply apps/explorer/lib/explorer/indexer/address_extraction.ex]: I'd be maybe OK with logging.  So, you're thinking just return no address if no pattern matches?

Comment (KronicDeth) [reply apps/explorer/lib/explorer/indexer/address_extraction.ex]: O, wait, thought of something why return no address doesn't work: if we return no address, the address won't get extracted, so the address will be missing from the address table as it won't get inserted before it is referenced by a foreign key in one of the other tables, so an error will occur and kill the task, but it will happen in the part of BlockFetcher that checks the `Chain.import_block` returns `{:ok, _}` instead of here, where we have more context.  Since we know the normal place this will called needs the extracted addresses or it will eventually fail, I think raising the exception closer to the fault makes more sense.

Does my reasoning make sense?

File: apps/explorer/lib/explorer/indexer/address_extraction.ex [hunk 5]
```diff
@@ -102,59 +315,146 @@ defmodule Explorer.Indexer.AddressExtraction do
...
     end
   end
 
-  defp merge_addresses(addresses) do
-    addresses
-    |> Enum.group_by(fn address -> address.hash end)
-    |> Enum.map(fn {_, similar_addresses} ->
-      Enum.reduce(similar_addresses, %{}, fn address, acc ->
-        Map.merge(acc, address)
-      end)
```
[the remark this answers is not in the document]
Comment (KronicDeth) [reply apps/explorer/lib/explorer/indexer/address_extraction.ex]: I'll give it a try.

[the remark this answers is not in the document]
Comment (KronicDeth) [reply apps/explorer/lib/explorer/indexer/address_extraction.ex]: Nope, that implementation is bugged.  It ends up combining results:

```
4) doctest Explorer.Indexer.AddressExtraction.extract_addresses/2 (2) (Explorer.Indexer.AddressExtractionTest)
     test/explorer/indexer/address_extraction_test.exs:6
     Doctest failed
     code: Explorer.Indexer.AddressExtraction.extract_addresses(
              %{
                internal_transactions: [
                  %{
                    block_number: 1,
                    from_address_hash: "0x0000000000000000000000000000000000000001"
                  },
                  %{
                    block_number: 2,
                    to_address_hash: "0x0000000000000000000000000000000000000002"
                  },
                  %{
                    block_number: 3,
                    created_contract_address_hash: "0x0000000000000000000000000000000000000003",
                    created_contract_code: "0x"
                  }
                ]
              }
            ) === [
             %{
               fetched_balance_block_number: 1,
               hash: "0x0000000000000000000000000000000000000001"
             },
             %{
               fetched_balance_block_number: 2,
               hash: "0x0000000000000000000000000000000000000002"
             },
             %{
               contract_code: "0x",
               fetched_balance_block_number: 3,
               hash: "0x0000000000000000000000000000000000000003"
             }
           ]
     left: [
             %{
               contract_code: "0x",
               fetched_balance_block_number: 3,
               hash: "0x0000000000000000000000000000000000000003"
             },
             %{
               contract_code: "0x",
               fetched_balance_block_number: 3,
               hash: "0x0000000000000000000000000000000000000003"
             },
             %{
               contract_code: "0x",
               fetched_balance_block_number: 3,
               hash: "0x0000000000000000000000000000000000000003"
             }
           ]
```

To see for yourself:

1. `git fetch origin`
2. `git checkout 'igorffs-#243'
3. `cd apps/explorer`
4. `MIX_ENV=test mix do ecto.drop, ecto.create, ecto.migrate`
5. `mix test test/explorer/indexer/address_extraction_test.exs`

It also produces `BadArityError` exceptions for `test/explorer/indexer/address_extraction_test.exs:149`, which is a test of completely empty data.

[the remark this answers is not in the document]
Comment (KronicDeth) [reply apps/explorer/lib/explorer/indexer/address_extraction.ex]: I am concerned that duping bug of `contract_code` and `hash` to all returned values is not something special about the change you gave me and could affect master.  I'll add tests on master to confirm.

[the remark this answers is not in the document]
Comment (igorffs) [reply apps/explorer/lib/explorer/indexer/address_extraction.ex]: Sorry @KronicDeth, I didn't get too much time to propose this, I was only trying to bring you the general idea. 

So there are a few bugs in this implementation beginning with the `max` function that should be `max_by` and where the `normalize_block_number` is placed at. If we use it at the end of the `merge_addresses` it will get the max number from the already merged items, so we should get it in the beginning like this:

```elixir
  defp merge_addresses(addresses) do
    addresses
    |> normalize_block_number()
    |> Enum.group_by(fn address -> address.hash end)
    |> Enum.map(fn {_, similar_addresses} ->
      Enum.reduce(similar_addresses, %{}, fn address, acc ->
        Map.merge(acc, address)
      end)
    end)
  end

  def normalize_block_number(addresses) do
    max_block_number =
      Enum.max_by(addresses, fn %{fetched_balance_block_number: num} -> num end)
      |> Map.get(:fetched_balance_block_number)


    Enum.map(addresses, fn item ->
      Map.merge(item, %{fetched_balance_block_number: max_block_number})
    end)
  end
```

That still may not be want you really want in case that we can only assign the max number in identical addresses and not all of the ones that appears in the `addresses` List. In that case, you'll probably need to use the `normalize_block_number()` after the group_by:

```elixir
  defp merge_addresses(addresses) do
    addresses
    |> Enum.group_by(fn address -> address.hash end)
    |> Enum.map(fn {_, similar_addresses} ->
      similar_addresses
      |> normalize_block_number()
      |> Enum.reduce(similar_addresses, %{}, fn address, acc ->
           Map.merge(acc, address)
         end)
    end)
  end
```

Which is your scenario?

[the remark this answers is not in the document]
Comment (KronicDeth) [reply apps/explorer/lib/explorer/indexer/address_extraction.ex]: The max number has to be per-address because we want that to be the a record of the last block in which an address is used.   I'll test that your new version does not assign the `contract_code` to all addresses in an entity group too on Monday.

[the remark this answers is not in the document]
Comment (KronicDeth) [reply apps/explorer/lib/explorer/indexer/address_extraction.ex]: The bug does not affect master.  So, that's fine, it is just the interaction of the merging with the changes you were suggesting.  Trying out your new suggestions now.

[the remark this answers is not in the document]
Comment (KronicDeth) [reply apps/explorer/lib/explorer/indexer/address_extraction.ex]:
```
addresses
    |> normalize_block_number()
```

is not the desired behavior.  We want the max block number that a specific address was used.  Using the max block across the group of addresses is no better than the timestamp as it doesn't tell us anything new about each address.  The goal for the `fetched_balance_block_number` to be that last time that we know an address was used for a balance transfer of any kind, so that collaborators or developer roles can track that no one's used their address when they weren't expecting it or that it was used when expecting it to for contracts that are triggered at a specific block number, which is a feature parity supports.

The second one, has the correct behavior.

[the remark this answers is not in the document]
Comment (KronicDeth) [reply apps/explorer/lib/explorer/indexer/address_extraction.ex]: `address_extract` still has a bug where it gets a `BadMapError` if the first extract does not match, but a later one does in the same extract format, such as when a block does not have a `number`, but it does have a `miner_hash`.

```elixir
test "blocks without a `number` aren't extracted" do
      assert Explorer.Indexer.AddressExtraction.extract_addresses(%{
               blocks: [
                 %{
                   miner_hash: "0xe8ddc5c7a2d2f0d7a9798459c0104fdf5e987aca"
                 }
               ]
             }) == []
    end
```

```
%BadMapError{term: []}
  5) test extract_addresses/1 blocks without a `number` aren't extracted (Explorer.Indexer.AddressExtractionTest)
     test/explorer/indexer/address_extraction_test.exs:19
     ** (BadMapError) expected a map, got: []
     code: assert Explorer.Indexer.AddressExtraction.extract_addresses(%{
     stacktrace:
       (stdlib) :maps.merge(%{hash: "0xe8ddc5c7a2d2f0d7a9798459c0104fdf5e987aca"}, [])
       (explorer) lib/explorer/indexer/address_extraction.ex:353: anonymous fn/3 in Explorer.Indexer.AddressExtraction.extract_address/2
       (elixir) lib/enum.ex:1899: Enum."-reduce/3-lists^foldl/2-0-"/3
       (elixir) lib/enum.ex:1294: Enum."-map/2-lists^map/1-0-"/2
       (elixir) lib/enum.ex:2924: Enum.flat_map_list/2
       (explorer) lib/explorer/indexer/address_extraction.ex:334: anonymous fn/3 in Explorer.Indexer.AddressExtraction.extract_addresses/2
       (stdlib) lists.erl:1263: :lists.foldl/3
       (explorer) lib/explorer/indexer/address_extraction.ex:332: Explorer.Indexer.AddressExtraction.extract_addresses/2
       test/explorer/indexer/address_extraction_test.exs:20: (test)
```

this is due to mixing lists and maps as the return type in `extract_address`.  This was one of the original concerns I had with this design as I would normally try to avoid mixing return types.  I'll attempt to produce a minimal diff fix.

[the remark this answers is not in the document]
Comment (KronicDeth) [reply apps/explorer/lib/explorer/indexer/address_extraction.ex]: Fixed in https://github.com/poanetwork/poa-explorer/commit/738f74a77b5139f2a56ba63c95b1135955a18af3.  Running full tests now to see that it working in integration tests.

[the remark this answers is not in the document]
Comment (KronicDeth) [reply apps/explorer/lib/explorer/indexer/address_extraction.ex]: Commit before force push, in case we want to recover error handling later: https://github.com/poanetwork/poa-explorer/blob/085f0d7e7428abff2bfe5e98394fa5363f09ada9/apps/explorer/lib/explorer/indexer/address_extraction.ex

**Comment by KronicDeth** [comment]
@igorffs your version of `AddressExtraction` with the minimum changes to make it pass the tests is now integrated.  Please re-review and Approve if no other changes are needed.

File: apps/explorer/test/explorer/indexer/address_extraction_test.exs [hunk 6]
```diff
@@ -55,116 +94,10 @@ defmodule Explorer.Indexer.AddressExtractionTest do
       assert Enum.empty?(addresses)
     end
 
-    test "addresses get merged when they're duplicated by their hash" do
```
Comment (igorffs) [inline apps/explorer/test/explorer/indexer/address_extraction_test.exs]: Can we bring the excluded tests back too since the code is back?
